lib/mpoolMonitor/mpool_monitor.go

```diff
@@ -0,0 +1,180 @@
+package mpoolMonitor
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/filecoin-project/boost/node/config"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/api/v1api"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+)
+
+var log = logging.Logger("mpoolmonitor")
+
+// timeStampedMsg wraps the pending msg with a chainEpoch
+type timeStampedMsg struct {
+	m     types.SignedMessage
+	added abi.ChainEpoch // Epoch when message was first noticed in mpool
+}
+
+type MpoolMonitor struct {
+	ctx                context.Context
+	cancel             context.CancelFunc
+	fullNode           v1api.FullNode
+	lk                 sync.Mutex
+	pendingAlertEpochs abi.ChainEpoch
+	msgs               map[cid.Cid]*timeStampedMsg
+}
+
+func NewMonitor(fullNode v1api.FullNode, cfg config.GraphqlConfig) *MpoolMonitor {
+	return &MpoolMonitor{
+		fullNode:           fullNode,
+		pendingAlertEpochs: abi.ChainEpoch(cfg.PendingAlertEpochs),
+		msgs:               make(map[cid.Cid]*timeStampedMsg),
+	}
+}
+
+func (mm *MpoolMonitor) Start(ctx context.Context) error {
+	log.Infow("Mpool monitor: starting")
+	mmctx, cancel := context.WithCancel(ctx)
+	mm.ctx = mmctx
+	mm.cancel = cancel
+	err := mm.update(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to start mpool monitor: %w", err)
+	}
+	go mm.startMonitoring(mmctx)
+	return nil
+}
+
+func (mm *MpoolMonitor) startMonitoring(ctx context.Context) {
+	ticker := time.NewTicker(10 * time.Second)
```

Comment:
I'd suggest running update once per epoch

```suggestion
	ticker := time.NewTicker(10 * time.Second)
```